pkg/util/file.go

```diff
@@ -261,28 +264,50 @@ func StreamDataToFile(r io.Reader, fileName string, preallocate bool) (int64, in
 
 type zeroWriterFunc func(*os.File, int64, int64) error
 
+func zeroWriterWithFallback(zwf zeroWriterFunc) func(dst *os.File, start, length int64) (int64, error) {
+	return func(dst *os.File, start, length int64) (int64, error) {
+		err := zwf(dst, start, length)
+		if err != nil {
+			klog.Errorf("Error zeroing range in destination file: %v, will write zeros directly", err)
+			err = AppendZeroWithWrite(dst, start, length)
+			if err != nil {
+				return 0, err
+			}
+			return length, nil
+		}
+		return 0, nil
+	}
+}
+
 func copyWithSparseCheck(dst *os.File, src io.Reader, zeroWriter zeroWriterFunc) (int64, int64, error) {
 	klog.Infof("copyWithSparseCheck to %s", dst.Name())
 	const buffSize = 32 * 1024
 	var bytesRead, bytesWritten int64
 	zeroBuf := make([]byte, buffSize)
 	writeBuf := make([]byte, buffSize)
 	var writeOffset int64
+	checkZeros := true
+	zeroWriterFunc := zeroWriterWithFallback(zeroWriter)
 	for {
 		nr, er := src.Read(writeBuf)
 		if nr > 0 {
 			var nw int
 			var ew error
-			if bytes.Equal(writeBuf[0:nr], zeroBuf[0:nr]) {
+			var zbw int64
+			if checkZeros && bytes.Equal(writeBuf[0:nr], zeroBuf[0:nr]) {
 				bytesRead += int64(nr)
 			} else {
 				if bytesRead > writeOffset {
-					// zeroWriter func should seek to bytesRead before returning
-					ew = zeroWriter(dst, writeOffset, bytesRead-writeOffset)
+					// func should seek to bytesRead before returning
+					zbw, ew = zeroWriterFunc(dst, writeOffset, bytesRead-writeOffset)
```

Comment:
I think using abbreviations like these (nw, we, er, etc.) makes the code hard to maintain, but the logic looks really good. Not a must so lgtm.